### PR TITLE
feat: add ERC721, transferable portfolio and shares

### PR DESCRIPTION
**Pull request by enitrat**
Solves #4, #9

Created the contract with two basic functions tested with foundry forge.

### CODE REVIEW

File: src/SharedFund.sol [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.13;
+
+import "openzeppelin-contracts/contracts/utils/Counters.sol";
+import "openzeppelin-contracts/contracts/token/ERC721/ERC721.sol";
+
+contract SharedFund is ERC721 {
+
+    using Counters for Counters.Counter;
+    Counters.Counter private _tokenIds;
```
Comment (dubloom) [inline src/SharedFund.sol]: to modify

**Review by dubloom — APPROVED** [review]

**Comment by patriciobcs** [comment]
@msaug you think you can refactor this PR to add the methods to the Portfolio? In other words: add ERC721 to the contract.

**Comment by enitrat** [comment]
I'll probably close this one and reopen a new, fresh one. Hopefully tomorrow afternoon

**Comment by patriciobcs** [comment]
Sure, sounds good, I created the issue #9

**Comment by enitrat** [comment]
ERC721 features, transferrable portfolio, deposit and withdraw features rebalancing users shares.
Added a pre-commit hook to format files with `forge fmt`
Install it with
```
chmod +xpre-commit.sh && cp pre-commit.sh .git/hooks/pre-commit.sh
```

I will create issues for the next actions to perform at this level

File: foundry.toml [hunk 2]
```diff
@@ -4,4 +4,5 @@ out = 'out'
 libs = ['lib']
 gas_reports = ["*"]
 optimizer = true
-optimizer_runs = 20000
\ No newline at end of file
+optimizer_runs = 20000
```
Comment (patriciobcs) [inline foundry.toml:7]: Is this necessary?

File: src/Portfolio.sol [hunk 3]
```diff
@@ -46,12 +54,100 @@ contract Portfolio is Ownable {
...
+     * @notice Deposit funds for a portfolio share. This rebalances the amount of shares each owner has.
+     * @param tokenId The ID of the token to deposit funds for.
+     */
+    function deposit(uint256 tokenId) public payable {
+        require(ownerOf(tokenId) == msg.sender, "You are not the owner of this share");
+        uint256 previousValue = getPortfolioValue();
+
+        // register the deposited ETH in the balances
+        uint256 depositedAmount = msg.value;
+        assets["ETH"].balance += depositedAmount;
```
Comment (patriciobcs) [inline src/Portfolio.sol]: Avoid using hardcoded values, this should allow any token

File: src/Portfolio.sol [hunk 4]
```diff
@@ -46,12 +54,100 @@ contract Portfolio is Ownable {
...
+        uint256 newValue = getPortfolioValue();
+
+        // update the shares by 1. rebalancing all shares and 2. overriding the current share with the new value
+        uint256 depositedValue = depositedAmount * uint256(priceFeeds.getLatestPrice("ETH"));
```
Comment (patriciobcs) [inline src/Portfolio.sol]: Same as last comment, no hardcoded values.

File: src/Portfolio.sol [hunk 5]
```diff
@@ -46,12 +54,100 @@ contract Portfolio is Ownable {
...
+        if (amount == PercentageMath.PERCENTAGE_FACTOR) {
+            newShareRebalanced = 0;
+        } else {
+            newShareRebalanced = (newShareValue * PercentageMath.PERCENTAGE_FACTOR) / (newTotalValue);
+        }
+        shares[tokenId] = newShareRebalanced;
+
+        // Convert USD values to ETH amount for withdrawal
+        uint256 etherToWithdraw = withdrawnValue / uint256(priceFeeds.getLatestPrice("ETH"));
+        emit log_uint(etherToWithdraw);
```
Comment (patriciobcs) [inline src/Portfolio.sol]: Remove unneeded logs

File: src/Portfolio.sol [hunk 6]
```diff
@@ -46,12 +54,100 @@ contract Portfolio is Ownable {
...
-    function addAsset(string memory _symbol, uint256 _balance, uint256 _proportion, bool _isFlexible, address _priceFeed) onlyOwner assetDoesNotExist(_symbol) public {
+    function withdraw(uint256 tokenId, uint256 amount) public {
+        require(ownerOf(tokenId) == msg.sender, "You are not the owner of this share");
+        uint256 share = shares[tokenId];
+        // TODO once we have swaps enabled, the withdrawn value should be calculated after slippage.
+
+        // USD Values
+        uint256 totalValue = getPortfolioValue();
+        uint256 shareValue = totalValue.percentMul(share);
+        uint256 withdrawnValue = totalValue.percentMul(share).percentMul(amount);
```
Comment (patriciobcs) [inline src/Portfolio.sol:117]: What if this value becomes greater than total value?

File: test/TestTransfer.t.sol [hunk 7]
```diff
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "./setup/TestSetup.sol";
+
+contract TestTransfer is TestSetup {
+    function testTransferShare() public {
+        deposit(user1, 1, 1 ether);
```
Comment (patriciobcs) [inline test/TestTransfer.t.sol:8]: deposit function is undeclared, test fail

File: test/setup/TestSetup.sol [hunk 8]
```diff
@@ -0,0 +1,45 @@
+import "../../src/Portfolio.sol";
```
Comment (patriciobcs) [inline test/setup/TestSetup.sol:4]: Why this file exists? It is duplicated and it is missing the license and solidity version

**Review by patriciobcs — CHANGES_REQUESTED** [review]
Extra observations: 
- Aave dependencies should be in `lib/` as a submodule and add the install to the Makefile as in the install command `forge install aave/aave-protocol`.
- Why did you remove the lib folder containing the submodules?
- Please add `forge test` to the pre-commit. Currently, there's a lot of test that don't run due to small details.

File: foundry.toml (hunk 2 above)
Comment (enitrat) [reply foundry.toml:7]: This was in your PR 😅

File: src/Portfolio.sol (hunk 3 above)
Comment (enitrat) [reply src/Portfolio.sol]: I would prefer that in an initial version, only ETH can be deposited and withdrawn from the contract.

File: test/setup/TestSetup.sol (hunk 8 above)
Comment (enitrat) [reply test/setup/TestSetup.sol:4]: For a modular testing architecture, we should use different test setups depending on what must be tested.
To test the deposit, withdraw, and share calculations functionalities, I want to set up a contract in this specific, basic state. 
If other tests need another contract state, they will override this setup.

**Comment by enitrat** [comment]
> Why did you remove the lib folder containing the submodules?

Should libraries be committed ? 🤔

File: foundry.toml (hunk 2 above)
Comment (patriciobcs) [reply foundry.toml:7]: Ups, Github trick me, I was referring to the next line (the comment)

File: test/setup/TestSetup.sol (hunk 8 above)
Comment (patriciobcs) [reply test/setup/TestSetup.sol:4]: But why the code is duplicated?

**Comment by patriciobcs** [comment]
> Should libraries be committed ? 🤔

I think that it is a good way to keep track of the version, what really matters. I didn't see other way of keeping a version in Foundry, the submodules is the way to go, they don't even use space in the repo, they are just references with a "commit".

File: src/Portfolio.sol (hunk 3 above)
Comment (patriciobcs) [reply src/Portfolio.sol]: Sounds reasonable, but let's keep the "ETH" string as a constant and no hardcoded.

File: test/setup/TestSetup.sol (hunk 8 above)
Comment (enitrat) [reply test/setup/TestSetup.sol:4]: wdym duplicated? where?

Comment (patriciobcs) [reply test/setup/TestSetup.sol:4]: With Portfolio.t.sol, Asset structure is duplicated and functions like setAsset also. I mean, we could change it later but it is better to have just one definition.

Comment (enitrat) [reply test/setup/TestSetup.sol:4]: I removed the duplicated code but have not yet refactored the `Portfolio.t.sol` file. Once merged, in the next PR we will refactor Portfolio.t.sol to isolate test cases by nature and create e2e tests.

**Review by patriciobcs — APPROVED** [review]

**Pull request merged**